### DISCUSSION TITLE
Flow retry functionality lacks parity with task retry functionality

### DISCUSSION THREAD

**Discussion by discdiver** [PrefectHQ/prefect#14621] [Ideas]
### First check

- [X] I added a descriptive title to this issue.
- [X] I used the GitHub search to find a similar request and didn't find it.
- [X] I searched the Prefect documentation for this feature.

### Describe the current behavior

Flows retries don't include a custom retry function option.

Retry delay options don't include a list of values, jitter, or exponential backoff.

This expands #9139 quite a bit, so closing that issue.

### Describe the proposed behavior

Example use of expanded retry delay behavior:

```python
@ flow (
retries=3,
retry_delay_seconds=exponential_backoff(backoff_factor=10),
retry_jitter_factor=1,
)
```

### Example Use

Better flexibility and better outcomes. Also less surprising behavior for flows if functionality matches tasks.

**Comment by cicdw** [comment]
I want to be careful with issues like this - having flows and tasks mirror each other isn't a worthy goal in and of itself; for example, what errors should a flow retry on that wouldn't be _better_ encapsulated and caught within a task? We need to have principled motivations for these interface extensions, not just matching the concepts for its own sake.  I'm not suggesting that some part of this issue isn't worthwhile, but the motivation is lacking to me.

**Comment by aaazzam** [comment]
I'm converting this issue to a discussion to allow for a more open-ended conversation about the proposed flow retry enhancements. As @cicdw pointed out, we need to explore more principled motivations and specific use cases for this functionality at the flow level.

We invite contributors to share:
- Detailed use cases where enhanced retry functionality (custom retry functions, jitter, exponential backoff) is necessary at the flow level rather than the task level.
- Examples demonstrating why this functionality can't be effectively handled by individual tasks.
- Any other insights or experiences related to flow-level retry needs.

This discussion will help us evaluate whether and how to implement these features. We look forward to a productive conversation!

**Comment by ricardogsilva** [comment]
I have a use case where I would like to be able to have a flow retry with exponential backoff and jitter.

I have a cluster with workers of type `process` where nodes have GPUs. I'm prototyping a way to manage GPU usage by leveraging prefect's builtin features. 
So I have setup a bunch of global concurrency locks for each GPU, with a capacity of one for each. Then at runtime I'm using prefect's [concurrency context manager](https://docs.prefect.io/v3/how-to-guides/workflows/global-concurrency-limits#using-the-concurrency-context-manager) to acquire a GPU-related limit. I have some logic for the flow run to try to acquire a free GPU on the worker and fail if none are available - then retry a bit later.

For better observability (and also for the sake of not overcomplicating my own code too much), I want to let prefect handle retry logic and I would like to use both exponential backoff and jitter.

I'd like to support two usage patterns:

1. A GPU is acquired when starting each task - I can use exponential backoff and jitter here;
2. A GPU is acquired per flow, and is used by all tasks - I can only use `retry_delay_seconds` here.
